lib/Dialect/OM/Transforms/VerifyObjectFields.cpp

```diff
@@ -0,0 +1,133 @@
+  auto &symbolTable = getAnalysis<SymbolTable>();
+
+  /// A map from a class and field name to a field.
+  llvm::MapVector<ClassLike, llvm::DenseMap<StringAttr, ClassFieldLike>> tables;
+  for (auto op : module->getRegion(0).getOps<om::ClassLike>())
+    tables.insert({op, llvm::DenseMap<StringAttr, ClassFieldLike>()});
+
+  // Peel tables parallelly.
+  mlir::parallelForEach(&getContext(), tables, [](auto &entry) {
+    ClassLike classLike = entry.first;
+    auto &table = entry.second;
+    classLike.walk([&](ClassFieldLike fieldLike) {
+      table.insert({fieldLike.getNameAttr(), fieldLike});
+    });
+  });
+
+  // Run actual verification. Make sure not to mutate `tables`.
+  auto result = mlir::failableParallelForEach(
+      &getContext(), tables, [&tables, &symbolTable](const auto &entry) {
+        ClassLike classLike = entry.first;
+        auto result =
+            classLike.walk([&](ObjectFieldOp objectField) -> WalkResult {
+              ObjectOp objectInst =
+                  objectField.getObject().getDefiningOp<ObjectOp>();
+              ClassLike classDef =
+                  cast<ClassLike>(symbolTable.lookupNearestSymbolFrom(
+                      objectField, objectInst.getClassNameAttr()));
+
+              // Traverse the field path, verifying each field exists.
+              ClassFieldLike finalField;
+              auto fields = SmallVector<FlatSymbolRefAttr>(
+                  objectField.getFieldPath().getAsRange<FlatSymbolRefAttr>());
+              for (size_t i = 0, e = fields.size(); i < e; ++i) {
+                // Verify the field exists on the ClassOp.
+                auto field = fields[i];
+                ClassFieldLike fieldDef;
+                auto it = tables.find(classDef);
+                assert(it != tables.end() && "must be vistied");
+                fieldDef = it->second.lookup(field.getAttr());
+
+                if (!fieldDef) {
+                  auto error =
+                      objectField.emitOpError("referenced non-existant field ")
+                      << field;
+                  error.attachNote(classDef.getLoc()) << "class defined here";
+                  return WalkResult::interrupt();
+                }
+
+                // If there are more fields, verify the current field is of
+                // ClassType, and look up the ClassOp for that field.
+                if (i < e - 1) {
+                  auto classType = dyn_cast<ClassType>(fieldDef.getType());
+                  if (!classType) {
+                    objectField.emitOpError("nested field access into ")
+                        << field << " requires a ClassType, but found "
+                        << fieldDef.getType();
+                    return WalkResult::interrupt();
+                  }
+
+                  // The nested ClassOp must exist, since a field with ClassType
+                  // must be an ObjectInstOp, which already verifies the class
+                  // exists.
```

Comment:
"me be an ObjectInstOp" -> that is, if this is valid and not an error, right?

If so, might be good to handle this either not finding a class or that the symbol resolves to a non-classlike.  I think `lookupNearestSymbolFrom<ClassLike>(..)` might help with that.

Reply:
This might not be needed, just checking 👍 .

Reply:
Yes I think it could crash when object is not created by ObjectInstOp. I added a check and test. Thanks!